src/main/java/org/folio/service/assignedusers/AssignedUsersServiceImpl.java

```diff
@@ -40,12 +48,53 @@
   private Converter<DbAssignedUser, AssignedUser> fromDbConverter;
   @Autowired
   private Converter<AssignedUser, User> userConverter;
+  @Autowired
+  private UsersLookUpService usersLookUpService;
 
+  @SneakyThrows
   @Override
   public CompletableFuture<AssignedUserCollection> findByCredentialsId(String credentialsId,
                                                                        Map<String, String> okapiHeaders) {
-    return assignedUserRepository.findByCredentialsId(toUUID(credentialsId), tenantId(okapiHeaders))
+    var dbAssignedUsers = assignedUserRepository.findByCredentialsId(toUUID(credentialsId), tenantId(okapiHeaders));
+
+    var ids = dbAssignedUsers
+      .thenApply(dbAssignedUsers1 -> dbAssignedUsers1.stream()
+        .map(DbAssignedUser::getId)
+        .collect(Collectors.toList()));
+
+    var users = new ArrayList<CompletableFuture<User>>();
+
+    ids.thenApply(idList -> idList.stream().map(id -> {
+      okapiHeaders.put(XOkapiHeaders.USER_ID, String.valueOf(id));
+      return users.add(usersLookUpService.lookUpUser(new OkapiParams(okapiHeaders)));
+    }));
+
+    var groupIds = users.stream()
+      .map(user -> user.thenApply(User::getPatronGroup)).collect(Collectors.toList());
+
+    var groups = new CompletableFuture<ArrayList<Group>>();
+
+    groupIds.forEach(groupId -> groupId.thenApply(id -> {
+      okapiHeaders.put(XOkapiHeaders.USER_ID, id);
+      return groups.thenApply(g -> {
+        try {
+          return g.add(usersLookUpService.lookUpGroup(new OkapiParams(okapiHeaders)).get());
```

Comment:
You shouldn't finalize futures. Just combine futures to get the expected result.